src/twisted/web/test/test_websocket.py

```diff
@@ -0,0 +1,344 @@
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from typing import Any, Generic, TypeVar
+from unittest import skipIf
+
+from zope.interface import implementer
+
+from twisted.internet.defer import Deferred
+from twisted.internet.error import ConnectionDone
+from twisted.internet.interfaces import IPushProducer
+from twisted.internet.protocol import Protocol, connectionDone
+from twisted.internet.testing import MemoryReactorClock
+from twisted.python.failure import Failure
+from twisted.test.iosim import ConnectionCompleter, IOPump
+from twisted.trial.unittest import SynchronousTestCase
+from twisted.web._responses import BAD_REQUEST
+from twisted.web.client import Agent, readBody
+from twisted.web.iweb import IRequest
+from twisted.web.resource import Resource
+from twisted.web.server import NOT_DONE_YET, Request, Site
+from twisted.web.static import Data
+
+WSP = TypeVar("WSP", bound="WebSocketProtocol")
+shouldSkip = False
+
+
+class WeirdResource(Resource):
+    def render_GET(self, request: Request) -> bytes:
+        """
+        Per U{the wsproto documentation
+        <https://python-hyper.org/projects/wsproto/en/latest/api.html#wsproto.events.RejectConnection.has_body>}:
+
+            - The only scenario in which the caller receives a RejectConnection
+              with C{has_body == False} is if the peer violates sends an
+              informational status code (1xx) other than 101
+
+        This is a weird edge case so we provoke it.
+        """
+        request.setResponseCode(102)
+        return b""
+
+
+class SlowResource(Resource):
+    def render_GET(self, request: Request) -> int:
+        self.request = request
+        return NOT_DONE_YET
+
+
+try:
+    __import__("wsproto")
+except ImportError:
+    shouldSkip = True
+else:
+    from twisted.web.websocket import (
+        ConnectionRejected,
+        WebSocketClientEndpoint,
+        WebSocketClientFactory,
+        WebSocketProtocol,
+        WebSocketResource,
+        WebSocketServerFactory,
+        WebSocketTransport,
+    )
+
+    @dataclass
+    class MyWSP(WebSocketProtocol):
```

Comment:
Having both server-side and client-side protocols implemented in the same class makes the code a bit harder to read.


```suggestion
    class MyWSP(WebSocketProtocol):
       """
       Used to implement both client-side and server-side WebSocket
       application that will help with testing. 
       """
```